### ISSUE TITLE
Document reason for why a task is skipped with onlyIf

### ISSUE THREAD

**Issue by big-guy** [gradle/gradle#3024]
### Expected Behavior

Build author can provide more context as to why a task is skipped.

### Current Behavior

When a task is skipped, but no reason is available.

### Context

For task output caching, we added a `cacheIf(String, Spec)` and `doNotCacheIf(String, Spec)`.  This allows plugins/build authors to specify a reason a task is not cacheable/is cacheable.

`onlyIf` is a similar API, but provides no way to provide additional context about why a task is skipped.

We could provide something like this for `onlyIf`:
```
task myTask(...) {
   onlyIf("today is sunday") {
      new Date().day == 0
   }
}
```

This would be useful to show in other reports.  We could say things like, "Task `myTask` was skipped because `myTask` can only execute if 'today is sunday'."

**Comment by JLLeitschuh** [comment]
This would be awesome for helping with debugging issues.

Would this only be reported with the `--info` flag or would it always be logged?

**Comment by lacasseio** [comment]
Most likely this information would only be display using `--info` flag. The lifecycle (default display) is only about the task outcome (skipped, cached, etc.). The info log gives informational information into why certain outcome came to be. Unfortunately, we aren't working on this at the moment so the behavior can't be confirmed.

**Comment by JLLeitschuh** [comment]
I'm guessing the behaviour would be the same as the `doNotCacheIf` in terms of when it would print.

**Comment by lacasseio** [comment]
You are right @JLLeitschuh, the behaviour should align with similar features.

**Comment by wolfs** [comment]
Fixed via https://github.com/gradle/gradle/pull/19428
